File: blink-macos/BlinkCore/Sources/BlinkCore/FlowDetection/FlowScoreCalculator.swift
import Foundation

/// Computes a composite flow score (0.0-1.0) from multiple input signals.
/// Updated every 30 seconds by the app's tick loop.
public final class FlowScoreCalculator {
    // Rolling buffers (timestamps in seconds since reference date)
    private var keystrokeTimestamps: [TimeInterval] = []
    private var mouseEvents: [MouseEvent] = []
    private var appSwitchTimestamps: [TimeInterval] = []
    private var titleChangeTimestamps: [TimeInterval] = []
    private var currentBundleID: String?

    // Scorers
    private let appSwitchScorer = AppSwitchScorer()
    private let keystrokeRhythmScorer = KeystrokeRhythmScorer()
    private let mouseBehaviorScorer = MouseBehaviorScorer()
    private let windowStabilityScorer = WindowStabilityScorer()
    private let contextBonusScorer = ContextBonusScorer()

    // Weights
    private static let weights: [(Double, String)] = [
        (0.35, "appSwitch"),
        (0.25, "keystroke"),
        (0.20, "mouse"),
        (0.10, "window"),
        (0.10, "context"),
    ]

    private static let maxBufferAge: TimeInterval = 600 // 10 minutes, prune older

    public init() {}

    // MARK: - Ingest events

    public func ingestKeystroke(_ event: KeystrokeEvent) {
        keystrokeTimestamps.append(event.timestamp)
    }

    public func ingestMouseEvent(_ event: MouseEvent) {
        mouseEvents.append(event)
    }

    public func recordAppSwitch(_ event: AppSwitchEvent) {
        appSwitchTimestamps.append(event.timestamp)
        currentBundleID = event.appBundleID
    }

    public func recordWindowTitleChange(at timestamp: TimeInterval) {
        titleChangeTimestamps.append(timestamp)
    }

    public func setCurrentApp(bundleID: String) {
        currentBundleID = bundleID
    }

    // MARK: - Compute score

    public func currentScore(now: TimeInterval) -> Double {
        pruneOldEvents(now: now)

        let keystrokeWindow: TimeInterval = 120
        let recentKeystrokeCount = keystrokeTimestamps.filter { $0 > now - keystrokeWindow }.count

        let appSwitchScore = appSwitchScorer.score(
            switchTimestamps: appSwitchTimestamps, now: now
        )
        let keystrokeScore = keystrokeRhythmScorer.score(
            keystrokeTimestamps: keystrokeTimestamps, now: now
        )
        let mouseScore = mouseBehaviorScorer.score(
            mouseEvents: mouseEvents, keystrokeCount: recentKeystrokeCount, now: now
        )
        let windowScore = windowStabilityScorer.score(
            titleChangeTimestamps: titleChangeTimestamps, now: now
        )
        let contextScore = contextBonusScorer.score(
            frontmostBundleID: currentBundleID
        )

        let scores = [appSwitchScore, keystrokeScore, mouseScore, windowScore, contextScore]
        let weightValues = Self.weights.map(\.0)

        return zip(weightValues, scores).map(*).reduce(0, +)
    }

    // MARK: - Maintenance

    public func reset() {
        keystrokeTimestamps.removeAll()
        mouseEvents.removeAll()
        appSwitchTimestamps.removeAll()
        titleChangeTimestamps.removeAll()
        currentBundleID = nil
    }

    private func pruneOldEvents(now: TimeInterval) {
        let cutoff = now - Self.maxBufferAge
        keystrokeTimestamps.removeAll { $0 < cutoff }
        mouseEvents.removeAll { $0.timestamp < cutoff }
        appSwitchTimestamps.removeAll { $0 < cutoff }
        titleChangeTimestamps.removeAll { $0 < cutoff }
    }
}

File: blink-windows/src/Blink.Core/FlowDetection/FlowScoreCalculator.cs
using Blink.Core.Abstractions;
using Blink.Core.FlowDetection.SignalProcessors;

namespace Blink.Core.FlowDetection;

public sealed class FlowScoreCalculator
{
    private readonly List<double> _keystrokeTimestamps = [];
    private readonly List<MouseEvent> _mouseEvents = [];
    private readonly List<double> _appSwitchTimestamps = [];
    private readonly List<double> _titleChangeTimestamps = [];
    private string? _currentAppId;

    private readonly AppSwitchScorer _appSwitchScorer = new();
    private readonly KeystrokeRhythmScorer _keystrokeRhythmScorer = new();
    private readonly MouseBehaviorScorer _mouseBehaviorScorer = new();
    private readonly WindowStabilityScorer _windowStabilityScorer = new();
    private readonly ContextBonusScorer _contextBonusScorer = new();

    private const double MaxBufferAge = 600; // 10 minutes

    // Weights
    private const double WeightAppSwitch = 0.35;
    private const double WeightKeystroke = 0.25;
    private const double WeightMouse = 0.20;
    private const double WeightWindow = 0.10;
    private const double WeightContext = 0.10;

    public void IngestKeystroke(KeystrokeEvent evt) => _keystrokeTimestamps.Add(evt.Timestamp);
    public void IngestMouseEvent(MouseEvent evt) => _mouseEvents.Add(evt);
    public void RecordAppSwitch(AppSwitchEvent evt)
    {
        _appSwitchTimestamps.Add(evt.Timestamp);
        _currentAppId = evt.AppId;
    }
    public void RecordWindowTitleChange(double timestamp) => _titleChangeTimestamps.Add(timestamp);
    public void SetCurrentApp(string appId) => _currentAppId = appId;

    public double CurrentScore(double now)
    {
        PruneOldEvents(now);

        const double keystrokeWindow = 120;
        var recentKeystrokeCount = _keystrokeTimestamps.Count(t => t > now - keystrokeWindow);

        var appSwitchScore = _appSwitchScorer.Score(_appSwitchTimestamps, now);
        var keystrokeScore = _keystrokeRhythmScorer.Score(_keystrokeTimestamps, now);
        var mouseScore = _mouseBehaviorScorer.Score(_mouseEvents, recentKeystrokeCount, now);
        var windowScore = _windowStabilityScorer.Score(_titleChangeTimestamps, now);
        var contextScore = _contextBonusScorer.Score(_currentAppId);

        return WeightAppSwitch * appSwitchScore
             + WeightKeystroke * keystrokeScore
             + WeightMouse * mouseScore
             + WeightWindow * windowScore
             + WeightContext * contextScore;
    }

    public void Reset()
    {
        _keystrokeTimestamps.Clear();
        _mouseEvents.Clear();
        _appSwitchTimestamps.Clear();
        _titleChangeTimestamps.Clear();
        _currentAppId = null;
    }

    private void PruneOldEvents(double now)
    {
        var cutoff = now - MaxBufferAge;
        _keystrokeTimestamps.RemoveAll(t => t < cutoff);
        _mouseEvents.RemoveAll(e => e.Timestamp < cutoff);
        _appSwitchTimestamps.RemoveAll(t => t < cutoff);
        _titleChangeTimestamps.RemoveAll(t => t < cutoff);
    }
}
